### ISSUE TITLE
Items.php -> General Search Service Provider System

### ISSUE THREAD

**Issue by robmyers** [creativecommons/garmonbozia#21]
Extract a base class from Items.php that can be subclassed for each search data provider.

The class provides methods to fetch and regularize data from a service.

It also identifies the kinds of media provided (image, text) and the range of available licenses.

Have a "providers" directory (better names gladly received) that provider classes live in.

e.g.: 

garmonbozia/providers/flickr.org.php

garmonbozia/providers/archive.org.php .

Each file *must* document the required additions to config.php to support its service:

e.g.:

"To use, get a key from:

https://www.flickr.com/services/api/misc.api_keys.html

and add to config.php as:

$flickr_api_key = '<YOUR_KEY>';"

Alternatives: json config in a garmonbozia/configs/ (pull request must provide a SERVICE_NAME-config.sample.php ), or a system that adds the config to config.php if you run the file directly. These both add overhead, so don't use them.

We don't want the system to scan the providers directory on every run through the database - require enabled providers to be explicitly listed in config.php. Combine this with the config system?

e.g.

$providers = [
  {'name' => 'flickr.com', 'media' => "iv", 'licenses' = 'bdnsz', 'api_key' => '<KEY>'},
  {'name' => 'archive.org', 'media' => "itsv", 'licenses' = 'bdnsz'},
];

Make sure it is easy for people to fork the repo, add a provider, and submit a Pull Request to incorporate it into master.

**Comment by robmyers** [comment]
Some providers are faster than others.

1. Run searches in parallel on their own pthreads.

2. Stream results to the user?

**Comment by robmyers** [comment]
Keep logs of searches that return nothing, and of service errors.

Keep tallies of successful, unsuccessful and failed searches in a database by hour. Use for alerting and reporting.

**Comment by rheaplex** [comment]
Moving the remaining elements to a new ticket.
